[lexical-list][lexical-markdown] Feature: Keep list marker for markdown round-trip

lytion opened the pull request:
## Description
I'm submitting a proposal so list markers are not loss anymore when switching from/to markdown.
This is a continuation in the idea that switching from/to markdown should not modify typed text.

These changes add a state to NodeList that save list marker between markdown import/export.

## Test plan
In lexical playground
In view mode:
1. Type `* hello`
2. Switch to markdown

### Before
-> markdown is `- hello`

### After
-> markdown is `* hello`, it kept marker
Switching again and again keep marker.

## Unit tests
Added unit tests in `LexicalMarkdown.test.ts` that verify a marker is kept between import/export
Added e2e tests in `Markdown.spec.mjs` that verify markers are kept when typing in lexical view and when copy/pasting.

etrepum commented on packages/lexical-markdown/src/MarkdownTransformers.ts:
```diff
@@ -210,6 +213,10 @@ const CODE_SINGLE_LINE_REGEX =
 const TABLE_ROW_REG_EXP = /^(?:\|)(.+)(?:\|)\s?$/;
 const TABLE_ROW_DIVIDER_REG_EXP = /^(\| ?:?-*:? ?)+\|\s?$/;
 
+export const listMarkerState = createState('listMarker', {
```
might be worth coming up with a more globally unique name for this, since it does only apply to markdown and it's not allowed for two different states on the same node to have the same name

lytion replied:
What about `mdListMarkerState` ?

etrepum commented on packages/lexical-markdown/src/MarkdownTransformers.ts:
```diff
@@ -251,6 +258,10 @@ const listReplace = (listType: ListType): ElementTransformer['replace'] => {
     );
+    if (listType === 'bullet' || listType === 'check') {
+      const marker = match[0].trim()[0];
+      $setState(listItem, listMarkerState, marker);
```
This state should really be on the list rather than the list item since it's not really permitted to mix different markers in the same unordered list. It's also a lot more compact as far as serialization goes.

if you put it on the ListItemNode what will happen is that you deserialize from markdown, then edit in lexical, you will end up with your custom markers and then any new items will be the default marker. You'll also run into the same sort of issues when copy and paste because the list items might have been created with different markers than their siblings.

Would be a good idea to have some test coverage that marks that

lytion commented on packages/lexical-markdown/src/MarkdownTransformers.ts:
```diff
@@ -258,11 +258,14 @@ const listReplace = (listType: ListType): ElementTransformer['replace'] => {
+        ? match[0].trim()[0]
+        : undefined;
     if ($isListNode(nextNode) && nextNode.getListType() === listType) {
+      if (listMarker) {
+        $setState(nextNode, listMarkerState, listMarker);
+      }
```
I'm not fully satisfied of this bunch of if condition but haven't found a better way to do it.

Except maybe putting `if ($isListNode(nextNode) && nextNode.getListType() === listType)` and the other if/else in variable to handle everything at the end. I didn't do that as it doesn't really simplify the code imo

etrepum commented on packages/lexical-markdown/src/MarkdownTransformers.ts:
```diff
@@ -210,6 +213,10 @@ const CODE_SINGLE_LINE_REGEX =
 const TABLE_ROW_REG_EXP = /^(?:\|)(.+)(?:\|)\s?$/;
 const TABLE_ROW_DIVIDER_REG_EXP = /^(\| ?:?-*:? ?)+\|\s?$/;
 
+export const listMarkerState = createState('listMarker', {
```
```suggestion
export const listMarkerState = createState('mdListMarker', {
```

lytion commented:
Done, I refactored the Unit Test part by:
>Added unit tests in LexicalMarkdown.test.ts that verify a marker is kept between import/export
Added e2e tests in Markdown.spec.mjs that verify markers are kept when typing in lexical view and when copy/pasting.

lytion commented:
It seems I'm missing something to merge it, I don't have the merge button.
I'm sorry I only started using GitHub PR system to work on Lexical, I'm not familiar with it. I can't find what's preventing me from merging

etrepum commented:
The merge button is only there for maintainers
